File: app/src/main/java/com/elewise/ldmobile/LoginActivity.java
package com.elewise.ldmobile;

import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.elewise.ldmobile.api.AuthStatusType;
import com.elewise.ldmobile.api.ParamAuthorizationRequest;
import com.elewise.ldmobile.api.ParamAuthorizationResponse;
import com.elewise.ldmobile.api.ParamTokenActivityCheckResponse;
import com.elewise.ldmobile.api.ResponseSessionActivityStatus;
import com.elewise.ldmobile.service.Prefs;
import com.elewise.ldmobile.service.Session;
import com.elewise.ldmobile.ui.DocsActivity;
import com.elewise.ldmobile.ui.SettingsActivity;
import com.elewise.ldmobile.utils.MessageUtils;

public class LoginActivity extends AppCompatActivity {

    private ProgressDialog progressDialog;
    private EditText etLogin;
    private EditText etPassword;
    private FloatingActionButton fbSettings;
    private AlertDialog dialog;
    private Session session;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        progressDialog = new ProgressDialog(this);
        progressDialog.setCancelable(false);
        progressDialog.setMessage("Вход в систему...");

        Button login = findViewById(R.id.btnLogin);
        etLogin = findViewById(R.id.etLogin);
        etPassword = findViewById(R.id.etPassword);
        fbSettings = findViewById(R.id.fbSettings);
        etLogin.setText(Prefs.INSTANCE.getLastLogin(this));

        session = Session.getInstance();

        login.setOnClickListener(view -> {
            if (TextUtils.isEmpty(etLogin.getText().toString())) {
                showError(getString(R.string.activity_login_enter_login));
                return;
            }

            if (TextUtils.isEmpty(etPassword.getText().toString())) {
                showError(getString(R.string.activity_login_enter_password));
                return;
            }

            runLogin(etLogin.getText().toString(), etPassword.getText().toString());
        });

        fbSettings.setOnClickListener(view -> {
            startActivity(new Intent(this, SettingsActivity.class));
        });

        getSupportActionBar().hide();

        // проверим активность сохраненной сессии, если есть
        if (!TextUtils.isEmpty(session.getToken())) {
            tokenActivityCheck(session.getToken());
        }
    }

    private void tokenActivityCheck(String token) {
        progressDialog.show();
        new Thread(() -> {
            ParamTokenActivityCheckResponse response = session.tokenActivityCheck(token);
            handleTokenActivityCheck(response);
        }).start();
    }

    private void runLogin(final String userName, final String password) {
        progressDialog.show();
        new Thread(() -> {
            ParamAuthorizationResponse response = session.getAuthToken(userName, password);
            handleLoginResponse(response);
        }).start();
    }

    private void handleTokenActivityCheck(final ParamTokenActivityCheckResponse response) {
        runOnUiThread(() -> {
            if (response != null) {
                // tood check activity_status null!! nullpointer
                if (response.getSession_activity_status().equals(ResponseSessionActivityStatus.Y.name())) {
                    // сессия еще активна, перейдем сразу к документам
                    openDocsActivity();
                } else {
                    if (!TextUtils.isEmpty(response.getMessage())) {
                        Toast.makeText(LoginActivity.this, response.getMessage(), Toast.LENGTH_LONG).show();
                    }
                    progressDialog.cancel();
                }
            } else {
                progressDialog.cancel();
            }
        });
    }

    private void handleLoginResponse(final ParamAuthorizationResponse response) {
        runOnUiThread(() -> {
            progressDialog.cancel();

            // вполнено с ошибками
            String errorMessage = getString(R.string.error_load_data);
            if (response != null) {
                if (response.getStatus().equals(AuthStatusType.S.name())) {
                    // успешно
                    session.setToken(response.getAccess_token());
                    Prefs.INSTANCE.saveLastLogin(LoginActivity.this, etLogin.getText().toString());
                    openDocsActivity();
                    return;
                } else {
                    if (!TextUtils.isEmpty(response.getMessage()))
                        errorMessage = response.getMessage();
                }
            }
            showError(errorMessage);
        });
    }

    private void showError(String errorMessage) {
        // показать ошибку
        dialog = MessageUtils.createDialog(LoginActivity.this, getString(R.string.alert_dialog_error), errorMessage);
        dialog.show();
    }

    private void openDocsActivity() {
        Intent intent = new Intent();
        intent.setClass(LoginActivity.this, DocsActivity.class);
        LoginActivity.this.startActivity(intent);
        LoginActivity.this.finish();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (progressDialog != null) progressDialog.dismiss();
        if (dialog != null) dialog.dismiss();
    }
}

File: app/src/main/java/com/elewise/ldmobile/ui/LoginActivity.kt
package com.elewise.ldmobile.ui

import android.app.AlertDialog
import android.app.ProgressDialog
import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.elewise.ldmobile.R
import com.elewise.ldmobile.api.AuthStatusType
import com.elewise.ldmobile.api.ParamAuthorizationResponse
import com.elewise.ldmobile.api.ParamTokenActivityCheckResponse
import com.elewise.ldmobile.api.ResponseSessionActivityStatus
import com.elewise.ldmobile.service.Prefs
import com.elewise.ldmobile.service.Session
import com.elewise.ldmobile.utils.MessageUtils
import kotlinx.android.synthetic.main.activity_login.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch


class LoginActivity : AppCompatActivity() {

    private var viewModelJob = Job()
    private val uiScope = CoroutineScope(Dispatchers.Main + viewModelJob)

    private var progressDialog: ProgressDialog? = null
    private var dialog: AlertDialog? = null
    private lateinit var session: Session

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)
        progressDialog = ProgressDialog(this).apply {
            setCancelable(false)
            setMessage(getString(R.string.login_progress))
        }
        val login = findViewById<Button>(R.id.btnLogin)
        etLogin.setText(Prefs.getLastLogin(this))
        session = Session.getInstance()
        login.setOnClickListener { view: View? ->
            if (Prefs.getConnectAddress(this).isEmpty()) {
                showError(getString(R.string.not_added_connection_address))
            }

            if (TextUtils.isEmpty(etLogin.getText().toString())) {
                showError(getString(R.string.activity_login_enter_login))
                return@setOnClickListener
            }
            if (TextUtils.isEmpty(etPassword.getText().toString())) {
                showError(getString(R.string.activity_login_enter_password))
                return@setOnClickListener
            }
            runLogin(etLogin.getText().toString(), etPassword.getText().toString())
        }
        fbSettings.setOnClickListener({ view: View? -> startActivity(Intent(this, SettingsActivity::class.java)) })
        supportActionBar?.hide()

        // проверим активность сохраненной сессии, если есть
        val lastToken = session.token
        if (!TextUtils.isEmpty(lastToken)) {
            tokenCheck(lastToken)
        }
    }

    private fun tokenCheck(token: String) {
        progressDialog?.show()
        uiScope.launch {
            try {
                val response = session.tokenActivityCheck(token).await()
                handleTokenActivityCheck(response.body())
            } catch (e: Exception) {
                progressDialog?.dismiss()
                Log.e("erorr token check", e.toString())
            }
        }
    }

    private fun runLogin(userName: String, password: String) {
        progressDialog?.show()
        uiScope.launch {
            try {
                val request = session.getAuthToken(userName, password)
                val response = request.await()
                handleLoginResponse(response.body())
            } catch (e: Exception) {
                progressDialog?.dismiss()
                showError(getString(R.string.error_load_data))
                Log.e("erorr login", e.toString())
            }
        }
    }

    private fun handleTokenActivityCheck(response: ParamTokenActivityCheckResponse?) {
        runOnUiThread {
            if (response != null) {
                if (response.session_activity_status != null && response.session_activity_status == ResponseSessionActivityStatus.Y.name) {
                    // сессия еще активна, перейдем сразу к документам
                    openDocsActivity()
                } else {
                    if (!TextUtils.isEmpty(response.message)) {
                        Toast.makeText(this@LoginActivity, response.message, Toast.LENGTH_LONG).show()
                    }
                    progressDialog?.cancel()
                }

                if (response.docs_size != null) {
                    session.docSize = response.docs_size
                }
            } else {
                progressDialog?.cancel()
            }
        }
    }

    private fun handleLoginResponse(response: ParamAuthorizationResponse?) {
        runOnUiThread {
            progressDialog?.cancel()

            // вполнено с ошибками
            var errorMessage: String? = getString(R.string.error_load_data)
            if (response != null) {
                if (response.status == AuthStatusType.S.name) {
                    // успешно
                    if (response.docs_size != null) {
                        session.docSize = response.docs_size
                    }

                    session.lastAuth = response
                    Prefs.saveLastLogin(this@LoginActivity, etLogin.text.toString())
                    openDocsActivity()
                    return@runOnUiThread
                } else {
                    if (!TextUtils.isEmpty(response.message)) errorMessage = response.message
                }
            }
            showError(errorMessage)
        }
    }

    private fun showError(errorMessage: String?) {
        // показать ошибку
        dialog = MessageUtils.createDialog(this@LoginActivity, getString(R.string.alert_dialog_error), errorMessage)
        dialog?.show()
    }

    private fun openDocsActivity() {
        val intent = Intent()
        intent.setClass(this@LoginActivity, DocsActivity::class.java)
        this@LoginActivity.startActivity(intent)
        finish()
    }

    override fun onDestroy() {
        super.onDestroy()
        viewModelJob.cancel()
        progressDialog?.dismiss()
        dialog?.dismiss()
    }
}